### PR TITLE
Delayed emit timing of `connect` event.

### PR DESCRIPTION
**Pull request by ogis-fujiwara**
Emit `connect` event after processing of `outgoingStore` is completed.

Problem:
When client publish new message during proccessing of publish/pubrel messages in `outgoingStore`, the newly published message interrupts the process. As a result, messages are not sent by published time order.

Outcomes:
Above message order problem is fixed by this commit. Added unit test verifies this case.

### CODE REVIEW

File: lib/client.js [hunk 1]
```diff
@@ -129,9 +129,12 @@ function MqttClient (streamBuilder, options) {
   // Inflight callbacks
   this.outgoing = {}
 
+  this.connectEmitter = null
+
   // Mark connected on connect
-  this.on('connect', function () {
+  this._onConnect = function () {
```
Comment (mcollina) [inline lib/client.js]: can you move this to a top level function instead?

File: lib/client.js [hunk 2]
```diff
@@ -141,6 +144,7 @@ function MqttClient (streamBuilder, options) {
     this.once('close', remove)
     outStore.on('end', function () {
       that.removeListener('close', remove)
+      that.connectEmitter()
```
Comment (mcollina) [inline lib/client.js]: can you just `emit('connect')`?

File: lib/client.js (hunk 1 above)
Comment (ogis-fujiwara) [reply lib/client.js]: Yes, I will do that.

**Comment by ogis-fujiwara** [comment]
I've updated all mentioned above.

Travis test was passed when I pushed changes to my topic branch. See below.
https://travis-ci.org/ogis-fujiwara/MQTT.js/builds/424621763
Would you restart failed tests?

File: lib/client.js [hunk 3]
```diff
@@ -915,7 +842,11 @@ MqttClient.prototype._handleConnack = function (packet) {
 
   if (rc === 0) {
     this.reconnecting = false
-    this.emit('connect', packet)
+    var that = this
+    this.connectEmitter = function () {
+      that.emit('connect', packet)
+    }
+    this._onConnect()
```
Comment (mcollina) [inline lib/client.js]: You can pass the packet to `_onConnect(packet)`  and wrap it in the closures there.

Comment (ogis-fujiwara) [reply lib/client.js]: Thanks, I will commit it.

**Comment by ogis-fujiwara** [comment]
I've commited it.
All Travis test against my branch was passed as below.
https://travis-ci.org/ogis-fujiwara/MQTT.js/builds/425060666

**Comment by RangerMauve** [comment]
I could see this changing assumptions made in existing codebases. +1 to it being semver-major.

File: lib/client.js [hunk 4]
```diff
@@ -129,74 +129,15 @@ function MqttClient (streamBuilder, options) {
...
-      } else if (outStore.destroy) {
-        outStore.destroy()
-      }
-    }
-
-    // start flowing
-    storeDeliver()
-  })
+  // True if connection is first time.
+  this.firstConnection = true
```
[the remark this answers is not in the document]
Comment (ogis-fujiwara) [reply lib/client.js]: I committed this. Could you please review?

**Comment by ogis-fujiwara** [comment]
Thank you very much for merging.

**Pull request merged**